### DISCUSSION TITLE
how best to cache shaper information?

### DISCUSSION THREAD

**Discussion by wez** [harfbuzz/harfbuzz#3856] [Q&A] [answered]
One of the hotspots in wezterm is the cost of shaping for busy terminal outputs.
I have some logic to try to cache and avoid re-shaping the same text at my application layer, but it is complex and if the text has changed I still need to call into harfbuzz.

https://harfbuzz.github.io/shaping-plans-and-caching.html talks about caching shaping plans, but it's not clear on exactly how to use them or when they might be invalidated or when best to reuse them, or what the gain might be from using them.

Could we have some more information about this, and other best practices to get the most performance out of harfbuzz?

Thanks!

**Comment by behdad** [comment, marked as answer]
The shape-plan caching happens automatically underneath `hb_shape`. So that wouldn't give you much.

I'll discuss other strategies in a separate comment.

**Comment by behdad** [comment]
> and other best practices to get the most performance out of harfbuzz?

I'm afraid you've hit the performance limit of HarfBuzz itself. From here on external (word?) caching might help.

**Comment by behdad** [comment]
Where does profiling show time is being spent? Disabling kerning / ligatures might help, since as a terminal you don't need them / discard them(?) anyway.

**Comment by behdad** [comment]
hb-ot-font has slightly better performance than hb-ft. But checking your code, I believe you've already tried that.
